### ISSUE TITLE
Harmonize sizing methods across components

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#227]
As @lindsaym-fa [points out](https://github.com/shoelace-style/webawesome/discussions/219#discussioncomment-11309343) size is an interesting one, as it’s both a qualitative (small/medium/large) as well as a quantitative parameter (`--size: 1.5em`). 

This is reflected in our own API as inconsistency: some components using a qualitative `size` attribute, others a `--size` CSS property, and others just responding to `font-size`:

Name |	Components
------|----------
`size` attribute | Button, Checkbox, Color Picker, Input, QR Code, Radio, Radio Button, Radio Group, Select, Switch,Tag, Textarea
`font-size` explicitly mentioned | [Icon](https://webawesome-git-next-font-awesome.vercel.app/docs/components/icon#sizing), [Icon Button](https://webawesome-git-next-font-awesome.vercel.app/docs/components/icon-button#sizes), [Spinner](https://webawesome-git-next-font-awesome.vercel.app/docs/components/spinner#size)
`--size` CSS property | [Avatar](https://webawesome-git-next-font-awesome.vercel.app/docs/components/avatar#css-custom-properties), [Drawer](https://webawesome-git-next-font-awesome.vercel.app/docs/components/drawer#css-custom-properties), [Progress Ring](https://webawesome-git-next-font-awesome.vercel.app/docs/components/progress-ring#size)
`--symbol-size` | [Rating](https://backers.webawesome.com/docs/components/rating#symbol-sizes)

I think we should pick a consistent scheme and go with it. It seems to me that both continuous sizing and token-based sizing are useful, token-based sizing to make common things easy, and continuous sizing as an escape hatch to make complex things possible. What if we do both? Something like:
- qualitative sizing via utility classes (`size-s`/`size-m`/`size-l`), which could resolve to different sizing tokens for different elements if needed (e.g. a large heading and a large button have vastly different font sizes), with some good defaults for arbitrary elements
- Continuous scaling via either `font-size` or a custom property that gets multiplied with other things to produce the final component `font-size`

Question: Are there components we want to keep a fixed size regardless of where they are placed or do we generally want things to scale based on context?

**Comment by claviska** [comment]
> Are there components we want to keep a fixed size regardless of where they are placed or do we generally want things to scale based on context?

In other explorations, I've found scaling with font size (using ems) has been pretty handy. It feels like the components just do the right thing most of the time, so I'd be in favor of that approach.

I still think it's valuable, from a design system perspective, to have something like "small/medium/large" that maps to specific sizes, but these could become thin wrappers around setting the font size.

One possible issue: we tend to set a specific height for buttons/form controls for consistency, because relying on paddings/line heights/etc. to get their has proven very hard to do (possibly due to rounding errors from using fractional values), so I'm not sure form controls can work the same way without solving that.

**Comment by LeaVerou** [comment]
Update: I’ve started some work on this, by introducing a [`size.css` helper](https://github.com/shoelace-style/webawesome/blob/next/src/styles/shadow/size.css) that components with a `size` attribute can use to simplify their CSS by defining variables with what the s, m, l sizes should be, and the helper then takes care of switching between them. It also defines utility classes like `.wa-size-s` that use the same variables, so that applied CSS can use this for natives too.

I've already implemented it in [`select`](https://github.com/shoelace-style/webawesome/commit/f9590195a5b9d43e4ba4c01460dfe1c877d6ee1f), [`textarea`](https://github.com/shoelace-style/webawesome/commit/53eb0d8f300004cf29bd5db771e47b05faa58078) and [`checkbox`](https://github.com/shoelace-style/webawesome/commit/9f0276f55eb71c9da51b6a90bb1e25f1f7c4f2ba).

**Comment by lindsaym-fa** [comment]
I was thinking about this more re: [Small/large version of dropdowns](https://github.com/shoelace-style/webawesome-alpha/discussions/139). I completely agree that supporting both utilities that can apply preset sizes (small | medium | large, whether via utility classes or `size` attribute — haven't quite processed the intersection with #231) and continuous sizing via `font-size` is valuable.

Refactoring our components to use `em` values for internal padding and margin rather than `--wa-space-*` variables seems like a needed step in this direction.

**Comment by LeaVerou** [comment]
I completely agree @lindsaym-fa! And a few days ago I even spotted `--symbol-size`, which 

So, concretely:
1. [ ] Drop `--size` and similar CSS properties, rework docs and CSS to just use `font-size`. This should be easy since these components are *already* using flexible padding.
2. [ ] Support `size` attributes on whatever element it makes sense, mention it in docs

I can make subissues for those if we have consensus.

> Refactoring our components to use `em` values for internal padding and margin rather than `--wa-space-*` variables seems like a needed step in this direction.

💯 
Though that's probably going to be trickier. It's not just about changing `rem` with `em`, right?

**Comment by lindsaym-fa** [comment]
Agreed with those concrete steps!

> Though that's probably going to be trickier. It's not just about changing rem with em, right?

Right, but with the right care and testing, it shouldn't be _too_ bad.
